src/tools/rbd_mirror/ImageSyncThrottler.cc

```diff
@@ -90,176 +79,85 @@ void ImageSyncThrottler<I>::start_sync(I *local_image_ctx, I *remote_image_ctx,
+           << dendl;
 
-    if (m_inflight_syncs.size() < m_max_concurrent_syncs &&
-        !m_sync_queue.empty()) {
-      next_sync_holder = m_sync_queue.back();
-      m_sync_queue.pop_back();
+  Mutex::Locker locker(m_lock);
 
-      assert(
-        m_inflight_syncs.count(next_sync_holder->m_local_image_id) == 0);
-      m_inflight_syncs[next_sync_holder->m_local_image_id] =
-        next_sync_holder;
-      dout(10) << "ready to start image sync for local_image_id "
-               << next_sync_holder->m_local_image_id << " ["
-               << m_inflight_syncs.size() << "/" << m_max_concurrent_syncs
-               << "]" << dendl;
-    }
+  auto it = m_waiting_syncs.find(sync_holder->m_local_image_id);
+  assert(it != m_waiting_syncs.end());
+  auto sync_holder_ctx = it->second;
+  m_waiting_syncs.erase(it);
 
-    dout(10) << "currently running image syncs [" << m_inflight_syncs.size()
-             << "/" << m_max_concurrent_syncs << "]" << dendl;
+  if (r == -ECANCELED) {
+    dout(10) << "canceled waiting image sync for local_image_id "
+             << sync_holder->m_local_image_id << dendl;
+    sync_holder_ctx->m_on_finish->complete(r);
+    sync_holder->m_sync->put();
+    delete sync_holder_ctx;
+    return;
   }
 
-  if (next_sync_holder) {
-    next_sync_holder->m_sync->send();
-  }
+  assert(r == 0);
+  m_inflight_syncs[sync_holder->m_local_image_id] = sync_holder_ctx;
+  sync_holder_ctx->m_sync->send();
 }
 
 template <typename I>
-void ImageSyncThrottler<I>::set_max_concurrent_syncs(uint32_t max) {
-  dout(20) << " max=" << max << dendl;
-
-  assert(max > 0);
-
-  std::list<C_SyncHolder *> next_sync_holders;
-  {
-    Mutex::Locker l(m_lock);
-    this->m_max_concurrent_syncs = max;
-
-    // Start waiting syncs in the case of available free slots
-    while(m_inflight_syncs.size() < m_max_concurrent_syncs
-          && !m_sync_queue.empty()) {
-        C_SyncHolder *next_sync_holder = m_sync_queue.back();
-        next_sync_holders.push_back(next_sync_holder);
-        m_sync_queue.pop_back();
-
-        assert(
-          m_inflight_syncs.count(next_sync_holder->m_local_image_id) == 0);
-        m_inflight_syncs[next_sync_holder->m_local_image_id] = next_sync_holder;
-
-        dout(10) << "ready to start image sync for local_image_id "
-                 << next_sync_holder->m_local_image_id << " ["
-                 << m_inflight_syncs.size() << "/" << m_max_concurrent_syncs
-                 << "]" << dendl;
-    }
-  }
+void ImageSyncThrottler<I>::handle_sync_finished(int r,
+                                                 C_SyncHolder *sync_holder) {
+  dout(20) << "local_image_id=" << sync_holder->m_local_image_id << ", r=" << r
+           << dendl;
 
-  for (const auto& sync_holder : next_sync_holders) {
-    sync_holder->m_sync->send();
-  }
+  sync_holder->m_on_finish->complete(r);
+  m_throttler->finish_op(sync_holder->m_local_image_id);
+  Mutex::Locker locker(m_lock);
```

Comment:
Nit: I would prefer to clean up the local state before firing the completion down to the instance throttler